wallet/udb/upgrades.go

```diff
@@ -489,6 +497,125 @@ func slip0044CoinTypeUpgrade(tx walletdb.ReadWriteTx, publicPassphrase []byte) e
 	return unifiedDBMetadata{}.putVersion(metadataBucket, newVersion)
 }
 
+func hasExpiryUpgrade(tx walletdb.ReadWriteTx, publicPassphrase []byte) error {
+	const oldVersion = 7
+	const newVersion = 8
+	metadataBucket := tx.ReadWriteBucket(unifiedDBMetadata{}.rootBucketKey())
+	txmgrBucket := tx.ReadWriteBucket(wtxmgrBucketKey)
+
+	// Assert this function is only called on version 7 databases.
+	dbVersion, err := unifiedDBMetadata{}.getVersion(metadataBucket)
+	if err != nil {
+		return err
+	}
+	if dbVersion != oldVersion {
+		const str = "hasExpiryUpgrade inappropriately called"
+		return apperrors.E{ErrorCode: apperrors.ErrUpgrade, Description: str, Err: nil}
+	}
+
+	fetchRawCreditScriptType := func(v []byte) scriptType {
+		if len(v) < creditValueSize {
+			return scriptTypeNonexisting
+		}
+		return scriptType(v[81] & ^accountExistsMask)
+	}
+
+	// Iterate through all mined credits and update stored serializations
+	creditsBucket := txmgrBucket.NestedReadWriteBucket(bucketCredits)
+	cursor := creditsBucket.ReadWriteCursor()
+	for k, v := cursor.First(); v != nil; k, v = cursor.Next() {
+		amt, change, err := fetchRawCreditAmountChange(v)
+		if err != nil {
+			return err
+		}
+
+		opCode := fetchRawCreditTagOpCode(v)
+		isCoinbase := fetchRawCreditIsCoinbase(v)
+		scrLoc := fetchRawCreditScriptOffset(v)
+		scrLen := fetchRawCreditScriptLength(v)
+		scrTy := fetchRawCreditScriptType(v)
+		account, err := fetchRawCreditAccount(v)
+		if err != nil {
+			return err
+		}
+
+		hash := extractRawCreditTxHash(k)
+		op := wire.OutPoint{
+			Hash:  hash,
+			Index: extractRawCreditIndex(k),
+		}
+
+		block, err := fetchBlockRecord(txmgrBucket, extractRawCreditHeight(k))
+		if err != nil {
+			return err
+		}
+
+		_, recV := existsTxRecord(txmgrBucket, &op.Hash, &block.Block)
+		record := &TxRecord{}
+		err = readRawTxRecord(&op.Hash, recV, record)
+		if err != nil {
+			return err
+		}
+
+		cred := &credit{
+			outPoint:   op,
+			block:      block.Block,
+			amount:     amt,
+			change:     change,
+			spentBy:    indexedIncidence{index: ^uint32(0)},
+			opCode:     opCode,
+			isCoinbase: isCoinbase,
+			hasExpiry:  record.MsgTx.Expiry != wire.NoExpiryValue,
+		}
+		updatedV := valueUnspentCredit(cred, scrTy, scrLoc, scrLen,
+			account, newVersion)
+		err = creditsBucket.Put(k, updatedV)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Iterate through all unmined credits and update stored serializations
+	unminedCreditsBucket := txmgrBucket.NestedReadWriteBucket(bucketUnminedCredits)
+	unminedCursor := unminedCreditsBucket.ReadWriteCursor()
+	for k, v := unminedCursor.First(); v != nil; k, v = unminedCursor.Next() {
+		amt, change, err := fetchRawUnminedCreditAmountChange(v)
```

Comment:
Instead of parsing out all of the data from the value, instead please copy it into a new byte slice and manually set the bit for the upgrade.  This is safer as it won't clobber other data in the credit value, such as spent-by information mapping it to a redeeming debit.